// @coinbase/cdp-agentkit-core/dist/actions/cdp/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferAction = exports.TradeAction = exports.RequestFaucetFundsAction = exports.RegisterBasenameAction = exports.MintNftAction = exports.GetBalanceAction = exports.DeployTokenAction = exports.DeployNftAction = exports.GetWalletDetailsAction = exports.CDP_ACTIONS = void 0;
exports.getAllCdpActions = getAllCdpActions;
const deploy_nft_1 = require("./deploy_nft");
Object.defineProperty(exports, "DeployNftAction", { enumerable: true, get: function () { return deploy_nft_1.DeployNftAction; } });
const deploy_token_1 = require("./deploy_token");
Object.defineProperty(exports, "DeployTokenAction", { enumerable: true, get: function () { return deploy_token_1.DeployTokenAction; } });
const get_balance_1 = require("./get_balance");
Object.defineProperty(exports, "GetBalanceAction", { enumerable: true, get: function () { return get_balance_1.GetBalanceAction; } });
const get_wallet_details_1 = require("./get_wallet_details");
Object.defineProperty(exports, "GetWalletDetailsAction", { enumerable: true, get: function () { return get_wallet_details_1.GetWalletDetailsAction; } });
const mint_nft_1 = require("./mint_nft");
Object.defineProperty(exports, "MintNftAction", { enumerable: true, get: function () { return mint_nft_1.MintNftAction; } });
const register_basename_1 = require("./register_basename");
Object.defineProperty(exports, "RegisterBasenameAction", { enumerable: true, get: function () { return register_basename_1.RegisterBasenameAction; } });
const request_faucet_funds_1 = require("./request_faucet_funds");
Object.defineProperty(exports, "RequestFaucetFundsAction", { enumerable: true, get: function () { return request_faucet_funds_1.RequestFaucetFundsAction; } });
const trade_1 = require("./trade");
Object.defineProperty(exports, "TradeAction", { enumerable: true, get: function () { return trade_1.TradeAction; } });
const transfer_1 = require("./transfer");
Object.defineProperty(exports, "TransferAction", { enumerable: true, get: function () { return transfer_1.TransferAction; } });
const wow_1 = require("./defi/wow");
/**
 * Retrieves all CDP action instances.
 * WARNING: All new CdpAction classes must be instantiated here to be discovered.
 *
 * @returns - Array of CDP action instances
 */
function getAllCdpActions() {
    return [
        new get_wallet_details_1.GetWalletDetailsAction(),
        new deploy_nft_1.DeployNftAction(),
        new deploy_token_1.DeployTokenAction(),
        new get_balance_1.GetBalanceAction(),
        new mint_nft_1.MintNftAction(),
        new register_basename_1.RegisterBasenameAction(),
        new request_faucet_funds_1.RequestFaucetFundsAction(),
        new trade_1.TradeAction(),
        new transfer_1.TransferAction(),
    ];
}
exports.CDP_ACTIONS = getAllCdpActions().concat(wow_1.WOW_ACTIONS);
