// @coinbase/cdp-agentkit-core/dist/actions/cdp/mint_nft.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MintNftAction = exports.MintNftInput = void 0;
exports.mintNft = mintNft;
const zod_1 = require("zod");
const MINT_NFT_PROMPT = `
This tool will mint an NFT (ERC-721) to a specified destination address onchain via a contract invocation. 
It takes the contract address of the NFT onchain and the destination address onchain that will receive the NFT as inputs. 
Do not use the contract address as the destination address. If you are unsure of the destination address, please ask the user before proceeding.
`;
/**
 * Input schema for mint NFT action.
 */
exports.MintNftInput = zod_1.z
    .object({
    contractAddress: zod_1.z.string().describe("The contract address of the NFT to mint"),
    destination: zod_1.z.string().describe("The destination address that will receive the NFT"),
})
    .strip()
    .describe("Instructions for minting an NFT");
/**
 * Mints an NFT (ERC-721) to a specified destination address onchain.
 *
 * @param wallet - The wallet to mint the NFT from.
 * @param args - The input arguments for the action.
 * @returns A message containing the NFT mint details.
 */
async function mintNft(wallet, args) {
    const mintArgs = {
        to: args.destination,
        quantity: "1",
    };
    try {
        const mintInvocation = await wallet.invokeContract({
            contractAddress: args.contractAddress,
            method: "mint",
            args: mintArgs,
        });
        const result = await mintInvocation.wait();
        return `Minted NFT from contract ${args.contractAddress} to address ${args.destination} on network ${wallet.getNetworkId()}.\nTransaction hash for the mint: ${result.getTransaction().getTransactionHash()}\nTransaction link for the mint: ${result.getTransaction().getTransactionLink()}`;
    }
    catch (error) {
        return `Error minting NFT: ${error}`;
    }
}
/**
 * Mint NFT action.
 */
class MintNftAction {
    constructor() {
        this.name = "mint_nft";
        this.description = MINT_NFT_PROMPT;
        this.argsSchema = exports.MintNftInput;
        this.func = mintNft;
    }
}
exports.MintNftAction = MintNftAction;
