// @coinbase/coinbase-sdk/dist/coinbase/smart_contract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartContract = void 0;
const api_1 = require("../client/api");
const transaction_1 = require("./transaction");
const types_1 = require("./types");
const coinbase_1 = require("./coinbase");
const utils_1 = require("./utils");
const errors_1 = require("./errors");
const contract_event_1 = require("./contract_event");
/**
 * A representation of a SmartContract on the blockchain.
 */
class SmartContract {
    /**
     * Creates a new SmartContract instance.
     *
     * @param contractModel - The SmartContract model from the API.
     */
    constructor(contractModel) {
        if (!contractModel) {
            throw new Error("SmartContract model cannot be empty");
        }
        this.model = contractModel;
    }
    /**
     * Returns a list of ContractEvents for the provided network, contract, and event details.
     *
     * @param networkId - The network ID.
     * @param protocolName - The protocol name.
     * @param contractAddress - The contract address.
     * @param contractName - The contract name.
     * @param eventName - The event name.
     * @param fromBlockHeight - The start block height.
     * @param toBlockHeight - The end block height.
     * @returns The contract events.
     */
    static async listEvents(networkId, protocolName, contractAddress, contractName, eventName, fromBlockHeight, toBlockHeight) {
        const contractEvents = [];
        const queue = [""];
        while (queue.length > 0) {
            const page = queue.shift();
            const response = await coinbase_1.Coinbase.apiClients.contractEvent.listContractEvents(networkId, protocolName, contractAddress, contractName, eventName, fromBlockHeight, toBlockHeight, page?.length ? page : undefined);
            response.data.data.forEach(contractEvent => {
                contractEvents.push(new contract_event_1.ContractEvent(contractEvent));
            });
            if (response.data.has_more) {
                if (response.data.next_page) {
                    queue.push(response.data.next_page);
                }
            }
        }
        return contractEvents;
    }
    /**
     * Converts a SmartContractModel into a SmartContract object.
     *
     * @param contractModel - The SmartContract model object.
     * @returns The SmartContract object.
     */
    static fromModel(contractModel) {
        return new SmartContract(contractModel);
    }
    /**
     * Returns the ID of the SmartContract.
     *
     * @returns The SmartContract ID.
     */
    getId() {
        return this.model.smart_contract_id;
    }
    /**
     * Returns the Network ID of the SmartContract.
     *
     * @returns The Network ID.
     */
    getNetworkId() {
        return this.model.network_id;
    }
    /**
     * Returns the Wallet ID that deployed the smart contract.
     *
     * @returns The Wallet ID.
     */
    getWalletId() {
        return this.model.wallet_id;
    }
    /**
     * Returns the Contract Address of the smart contract.
     *
     * @returns The Contract Address.
     */
    getContractAddress() {
        return this.model.contract_address;
    }
    /**
     * Returns the Deployer Address of the smart contract.
     *
     * @returns The Deployer Address.
     */
    getDeployerAddress() {
        return this.model.deployer_address;
    }
    /**
     * Returns the Type of the smart contract.
     *
     * @returns The Smart Contract Type.
     */
    getType() {
        switch (this.model.type) {
            case api_1.SmartContractType.Erc20:
                return types_1.SmartContractType.ERC20;
            case api_1.SmartContractType.Erc721:
                return types_1.SmartContractType.ERC721;
            case api_1.SmartContractType.Erc1155:
                return types_1.SmartContractType.ERC1155;
            default:
                throw new Error(`Unknown smart contract type: ${this.model.type}`);
        }
    }
    /**
     * Returns the Options of the smart contract.
     *
     * @returns The Smart Contract Options.
     */
    getOptions() {
        if (this.isERC20(this.getType(), this.model.options)) {
            return {
                name: this.model.options.name,
                symbol: this.model.options.symbol,
                totalSupply: this.model.options.total_supply,
            };
        }
        else if (this.isERC721(this.getType(), this.model.options)) {
            return {
                name: this.model.options.name,
                symbol: this.model.options.symbol,
                baseURI: this.model.options.base_uri,
            };
        }
        else {
            return {
                uri: this.model.options.uri,
            };
        }
    }
    /**
     * Returns the ABI of the smart contract.
     *
     * @returns The ABI as a JSON-encoded string.
     */
    getAbi() {
        return JSON.parse(this.model.abi);
    }
    /**
     * Returns the Transaction of the smart contract deployment.
     *
     * @returns The Transaction.
     */
    getTransaction() {
        return new transaction_1.Transaction(this.model.transaction);
    }
    /**
     * Signs the SmartContract deployment with the provided key and returns the hex signature
     * required for broadcasting the SmartContract deployment.
     *
     * @param key - The key to sign the SmartContract deployment with
     * @returns The hex-encoded signed payload
     */
    async sign(key) {
        return this.getTransaction().sign(key);
    }
    /**
     * Broadcasts the SmartContract deployment to the Network.
     *
     * @returns The SmartContract object
     * @throws {APIError} if the API request to broadcast a SmartContract deployment fails.
     */
    async broadcast() {
        if (!this.getTransaction().isSigned())
            throw new Error("Cannot broadcast unsigned SmartContract deployment");
        const deploySmartContractRequest = {
            signed_payload: this.getTransaction().getSignature(),
        };
        const response = await coinbase_1.Coinbase.apiClients.smartContract.deploySmartContract(this.getWalletId(), this.getDeployerAddress(), this.getId(), deploySmartContractRequest);
        return SmartContract.fromModel(response.data);
    }
    /**
     * Waits for the SmartContract deployment to be confirmed on the Network or fail on chain.
     * Waits until the SmartContract deployment is completed or failed on-chain by polling at the given interval.
     * Raises an error if the SmartContract deployment takes longer than the given timeout.
     *
     * @param options - The options to configure the wait function.
     * @param options.intervalSeconds - The interval to check the status of the SmartContract deployment.
     * @param options.timeoutSeconds - The maximum time to wait for the SmartContract deployment to be confirmed.
     *
     * @returns The SmartContract object in a terminal state.
     * @throws {Error} if the SmartContract deployment times out.
     */
    async wait({ intervalSeconds = 0.2, timeoutSeconds = 10 } = {}) {
        const startTime = Date.now();
        while (Date.now() - startTime < timeoutSeconds * 1000) {
            await this.reload();
            // If the SmartContract deployment is in a terminal state, return the SmartContract.
            const status = this.getTransaction().getStatus();
            if (status === types_1.TransactionStatus.COMPLETE || status === types_1.TransactionStatus.FAILED) {
                return this;
            }
            await (0, utils_1.delay)(intervalSeconds);
        }
        throw new errors_1.TimeoutError("SmartContract deployment timed out");
    }
    /**
     * Reloads the SmartContract model with the latest data from the server.
     *
     * @throws {APIError} if the API request to get a SmartContract fails.
     */
    async reload() {
        const result = await coinbase_1.Coinbase.apiClients.smartContract.getSmartContract(this.getWalletId(), this.getDeployerAddress(), this.getId());
        this.model = result?.data;
    }
    /**
     * Returns a string representation of the SmartContract.
     *
     * @returns The string representation of the SmartContract.
     */
    toString() {
        return (`SmartContract{id: '${this.getId()}', networkId: '${this.getNetworkId()}', ` +
            `contractAddress: '${this.getContractAddress()}', deployerAddress: '${this.getDeployerAddress()}', ` +
            `type: '${this.getType()}'}`);
    }
    /**
     * Type guard for checking if the smart contract is an ERC20.
     *
     * @param type - The type of the smart contract.
     * @param options - The options of the smart contract.
     * @returns True if the smart contract is an ERC20, false otherwise.
     */
    isERC20(type, options) {
        return type === types_1.SmartContractType.ERC20;
    }
    /**
     * Type guard for checking if the smart contract is an ERC721.
     *
     * @param type - The type of the smart contract.
     * @param options - The options of the smart contract.
     * @returns True if the smart contract is an ERC721, false otherwise.
     */
    isERC721(type, options) {
        return type === types_1.SmartContractType.ERC721;
    }
}
exports.SmartContract = SmartContract;
