// @coinbase/coinbase-sdk/dist/coinbase/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartContractType = exports.StakeOptionsMode = exports.ServerSignerStatus = exports.FundOperationStatus = exports.PayloadSignatureStatus = exports.StakingRewardFormat = exports.ValidatorStatus = exports.SponsoredSendStatus = exports.TransactionStatus = exports.TransferStatus = void 0;
/**
 * Transfer status type definition.
 */
var TransferStatus;
(function (TransferStatus) {
    TransferStatus["PENDING"] = "pending";
    TransferStatus["BROADCAST"] = "broadcast";
    TransferStatus["COMPLETE"] = "complete";
    TransferStatus["FAILED"] = "failed";
})(TransferStatus || (exports.TransferStatus = TransferStatus = {}));
/**
 * Transaction status type definition.
 */
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus["PENDING"] = "pending";
    TransactionStatus["SIGNED"] = "signed";
    TransactionStatus["BROADCAST"] = "broadcast";
    TransactionStatus["COMPLETE"] = "complete";
    TransactionStatus["FAILED"] = "failed";
    TransactionStatus["UNSPECIFIED"] = "unspecified";
})(TransactionStatus || (exports.TransactionStatus = TransactionStatus = {}));
/**
 * Sponsored Send status type definition.
 */
var SponsoredSendStatus;
(function (SponsoredSendStatus) {
    SponsoredSendStatus["PENDING"] = "pending";
    SponsoredSendStatus["SIGNED"] = "signed";
    SponsoredSendStatus["SUBMITTED"] = "submitted";
    SponsoredSendStatus["COMPLETE"] = "complete";
    SponsoredSendStatus["FAILED"] = "failed";
})(SponsoredSendStatus || (exports.SponsoredSendStatus = SponsoredSendStatus = {}));
/**
 * Validator status type definition.
 * Represents the various states a validator can be in.
 */
var ValidatorStatus;
(function (ValidatorStatus) {
    ValidatorStatus["UNKNOWN"] = "unknown";
    ValidatorStatus["PROVISIONING"] = "provisioning";
    ValidatorStatus["PROVISIONED"] = "provisioned";
    ValidatorStatus["DEPOSITED"] = "deposited";
    ValidatorStatus["PENDING_ACTIVATION"] = "pending_activation";
    ValidatorStatus["ACTIVE"] = "active";
    ValidatorStatus["EXITING"] = "exiting";
    ValidatorStatus["EXITED"] = "exited";
    ValidatorStatus["WITHDRAWAL_AVAILABLE"] = "withdrawal_available";
    ValidatorStatus["WITHDRAWAL_COMPLETE"] = "withdrawal_complete";
    ValidatorStatus["ACTIVE_SLASHED"] = "active_slashed";
    ValidatorStatus["EXITED_SLASHED"] = "exited_slashed";
    ValidatorStatus["REAPED"] = "reaped";
})(ValidatorStatus || (exports.ValidatorStatus = ValidatorStatus = {}));
/**
 * Staking reward format type definition.
 * Represents the format in which staking rewards can be queried.
 */
var StakingRewardFormat;
(function (StakingRewardFormat) {
    StakingRewardFormat["USD"] = "usd";
    StakingRewardFormat["NATIVE"] = "native";
})(StakingRewardFormat || (exports.StakingRewardFormat = StakingRewardFormat = {}));
/**
 * Payload Signature status type definition.
 */
var PayloadSignatureStatus;
(function (PayloadSignatureStatus) {
    PayloadSignatureStatus["PENDING"] = "pending";
    PayloadSignatureStatus["SIGNED"] = "signed";
    PayloadSignatureStatus["FAILED"] = "failed";
})(PayloadSignatureStatus || (exports.PayloadSignatureStatus = PayloadSignatureStatus = {}));
/**
 * Fund Operation status type definition.
 */
var FundOperationStatus;
(function (FundOperationStatus) {
    FundOperationStatus["PENDING"] = "pending";
    FundOperationStatus["COMPLETE"] = "complete";
    FundOperationStatus["FAILED"] = "failed";
})(FundOperationStatus || (exports.FundOperationStatus = FundOperationStatus = {}));
/**
 * ServerSigner status type definition.
 */
var ServerSignerStatus;
(function (ServerSignerStatus) {
    ServerSignerStatus["PENDING"] = "pending_seed_creation";
    ServerSignerStatus["ACTIVE"] = "active_seed";
})(ServerSignerStatus || (exports.ServerSignerStatus = ServerSignerStatus = {}));
/**
 * StakeOptionsMode type definition.
 */
var StakeOptionsMode;
(function (StakeOptionsMode) {
    /**
     * Defaults to the mode specific to the asset.
     */
    StakeOptionsMode["DEFAULT"] = "default";
    /**
     * Partial represents Partial Ethereum Staking mode.
     */
    StakeOptionsMode["PARTIAL"] = "partial";
    /**
     * Native represents Native Ethereum Staking mode.
     */
    StakeOptionsMode["NATIVE"] = "native";
})(StakeOptionsMode || (exports.StakeOptionsMode = StakeOptionsMode = {}));
/**
 * Smart Contract Type
 */
var SmartContractType;
(function (SmartContractType) {
    SmartContractType["ERC20"] = "erc20";
    SmartContractType["ERC721"] = "erc721";
    SmartContractType["ERC1155"] = "erc1155";
})(SmartContractType || (exports.SmartContractType = SmartContractType = {}));
